### ISSUE TITLE
auto derivation of validator for sealed trait with Magnolia gives unexpected results

### ISSUE THREAD

**Issue by 0lejk4** [softwaremill/tapir#363]
So I discovered one interesting moment with Validator derivation: my endpoint returns sealed trait, 
and there is Validator for some fields of both coproducts if I return in my endpoints just case classes of coproducts -  validation is shown in OpenAPI as expected, but if I use sealed trait as return value both coproducts lose all validation rules from a schema definition. But there is one interesting moment, I have a list of ServerEndpoints with such sealed trait as the return type and if I change just the first ServerEndpoint in this list to return just case class and others are left unchanged generated OpenAPI looks like expected with all validators. :face_with_head_bandage: 
Still haven`t found out why this happens but this behaviour is really annoying.

**Comment by adamw** [comment]
Not sure I understand - could you post a code example reproducing the problem?

**Comment by 0lejk4** [comment]
Here is the example:
```
object Test extends App {
  import io.circe.generic.extras.semiauto._
  implicit val configuration: Configuration = Configuration.default.withDiscriminator("type")

  sealed trait A
  object A {
    implicit val codec: Codec[A] = deriveConfiguredCodec
  }
  case class B(name: Name) extends A
  object B {
    implicit val codec: Codec[B] = deriveConfiguredCodec
  }

  object Names extends Enumeration {
    type Name = Value
    val Oleg = Value("OLEG")
    val Sanya = Value("SANYA")
    implicit val codec: Codec[Name] = Codec.codecForEnumeration(this)
    implicit val enum: Validator[Name] = enumValidator[Name]
  }

  val endpoints: List[ServerEndpoint[_, _, _, EntityBody[Effect], Effect]] = List(endpoint1, endpoint2)
  def endpoint1 = commonEndpoint("", "").out(jsonBody[A]).effectServerLogic(_ => ZIO.succeed[A](B(Names.Oleg)))
  def endpoint2 = commonEndpoint("", "").out(jsonBody[A]).effectServerLogic(_ => ZIO.succeed[A](B(Names.Oleg)))

  println(endpoints.toOpenAPI("", "").toYaml)
}
```

In this case B scheme is 
B:
      allOf:
      - $ref: '#/components/schemas/A'
      - required:
        - name
        type: object
        properties:
          name:
            type: string

but if I change endpoint1 to return B like so:
`  def endpoint1 = commonEndpoint("", "").out(jsonBody[B]).effectServerLogic(_ => ZIO.succeed[B](B(Names.Oleg)))
`
B will get correct scheme - B:
      allOf:
      - $ref: '#/components/schemas/A'
      - required:
        - name
        type: object
        properties:
          name:
            type: string
            enum:
            - OLEG
            - SANYA

**Comment by 0lejk4** [comment]
Also when I debug I see that in first case validators are not applied, while in the second case they are. Don`t mind `commonEndpoint`, `effectServerLogic` and `enumValidator` those are my custom definitions. Also, Tapir version is the latest. You can try to reproduce it, this is relatively easy.

**Comment by adamw** [comment]
Ok, I'll try, though it would be easiest if you would just give me code which I could copy-paste, run and reproduce the problem :) (even if it's easy to do yourself ;) )

**Comment by 0lejk4** [comment]
```
object Problem extends App {
  import Names.Name
  import cats.effect.IO
  import org.http4s.EntityBody
  import io.circe.Codec
  import io.circe.generic.extras.Configuration
  import io.circe.generic.extras.semiauto._
  import sttp.tapir.server.ServerEndpoint
  import sttp.tapir.Validator
  import sttp.tapir.json.circe._
  import sttp.tapir._
  import sttp.tapir.docs.openapi._
  import sttp.tapir.openapi.circe.yaml._

  implicit val configuration: Configuration = Configuration.default.withDiscriminator("type")

  object Names extends Enumeration {
    type Name = Value
    val Oleg = Value("OLEG")
    val Sanya = Value("SANYA")
    implicit val codec: Codec[Name] = Codec.codecForEnumeration(this)
    implicit val validator: Validator[Name] = Validator.enum(List(Oleg, Sanya)).encode(_.toString)
    implicit val schema: Schema[Name] = Schema(SchemaType.SString)
  }

  sealed trait A
  object A {
    implicit val codec: Codec[A] = deriveConfiguredCodec
  }
  case class B(name: Name, string: String) extends A
  object B {
    implicit val validator: Validator[String] = Validator.minLength(1) and Validator.maxLength(255)
    implicit val codec: Codec[B] = deriveConfiguredCodec
  }


  val endpoints: List[ServerEndpoint[_, _, _, EntityBody[IO], IO]] = List(endpoint1, endpoint2)
  def endpoint1 = endpoint.get.out(jsonBody[A]).serverLogic[IO](_ => IO.apply[A](B(Names.Oleg, "23")).map(Right(_)))
  def endpoint2 = endpoint.post.out(jsonBody[A]).serverLogic[IO](_ => IO.apply[A](B(Names.Oleg, "23")).map(Right(_)))

  println(endpoints.toOpenAPI("", "").toYaml)
}

```
As I said change jsonBody[A] to be jsonBody[B] in endpoint1 and everything will work.

**Comment by adamw** [comment]
I see where the problem is. In `ValidatorMagnoliaDerivation`, we don't handle coproducts:

```
def dispatch[T](ctx: SealedTrait[Validator, T]): Validator[T] = Validator.pass
```

We should generate a parallel structure for validators, as for schemas. Not sure what the problems are here, maybe @ghostbuster91 will have some more insights. However, one problem I can see is that we'll need to traverse both structures in parallel, when generating the docs - `Schema` and `Validator`.

Maybe it's a good idea to combine `Schema` and `Validator` after all? But that would be a bigger change.

Sorry, but no quick fixes for this problem for now

**Comment by adamw** [comment]
Thanks for the example though, this allowed me diagnose the problem quickly :)

**Comment by adamw** [comment]
Fixed in 0.12.13
